agent/agent.go

```diff
@@ -205,6 +209,154 @@ func (a *agent) runLoop(ctx context.Context) {
+	err = cmd.Run()
+
+	const bufLimit = 10 << 10
+	if out.Len() > bufLimit {
+		err = errors.Join(
+			err,
+			xerrors.Errorf("output truncated from %v to %v bytes", out.Len(), bufLimit),
+		)
+		out.Truncate(bufLimit)
+	}
+
+	if err != nil {
+		result.Error = err.Error()
+	}
+	result.Value = out.String()
+	return result
+}
+
+func convertInterval(i int64) time.Duration {
+	// In tests we want to set shorter intervals because engineers are
+	// impatient.
+	base := time.Second
+	if flag.Lookup("test.v") != nil {
+		base = time.Millisecond * 100
+	}
+	return time.Duration(i) * base
+}
+
+type metadataResultAndKey struct {
+	result *codersdk.WorkspaceAgentMetadataResult
+	key    string
+}
+
+func (a *agent) reportMetadataLoop(ctx context.Context) {
+	baseInterval := convertInterval(1)
+
+	var (
+		baseTicker       = time.NewTicker(baseInterval)
+		lastCollectedAts = make(map[string]time.Time)
+		metadataResults  = make(chan metadataResultAndKey, 16)
+	)
+	defer baseTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case mr := <-metadataResults:
+			lastCollectedAts[mr.key] = mr.result.CollectedAt
+			err := a.client.PostMetadata(ctx, mr.key, *mr.result)
+			if err != nil {
+				a.logger.Error(ctx, "report metadata", slog.Error(err))
+			}
+		case <-baseTicker.C:
+			break
+		}
+
+		if len(metadataResults) > cap(metadataResults)/2 {
+			// If we're backpressured on sending back results, we risk
+			// runaway goroutine growth and/or overloading coderd. So,
+			// we just skip the collection. Since we never update
+			// the collections map, we'll retry the collection
+			// on the next tick.
+			a.logger.Debug(
+				ctx, "metadata collection backpressured",
+				slog.F("queue_len", len(metadataResults)),
+			)
+			continue
+		}
+
+		manifest := a.manifest.Load()
+		if manifest == nil {
+			continue
+		}
+		// If the manifest changes (e.g. on agent reconnect) we need to
+		// purge old cache values to prevent lastCollectedAt from growing
+		// boundlessly.
+		for key := range lastCollectedAts {
+			if slices.IndexFunc(manifest.Metadata, func(md codersdk.WorkspaceAgentMetadataDescription) bool {
+				return md.Key == key
+			}) < 0 {
+				delete(lastCollectedAts, key)
+			}
+		}
+
+		// Spawn a goroutine for each metadata collection, and use a
+		// channel to synchronize the results and avoid both messy
+		// mutex logic and overloading the API.
+		for _, md := range manifest.Metadata {
+			collectedAt, ok := lastCollectedAts[md.Key]
+			if ok {
+				// If the interval is zero, we assume the user just wants
+				// a single collection at startup, not a spinning loop.
+				if md.Interval == 0 {
+					continue
+				}
+				if collectedAt.Add(
+					convertInterval(md.Interval),
+				).After(time.Now()) {
+					continue
+				}
+			}
+
+			go func(md codersdk.WorkspaceAgentMetadataDescription) {
+				select {
+				case <-ctx.Done():
+					return
+				case metadataResults <- metadataResultAndKey{
```

Comment:
This lacks a default, so if I have 16 metadata on an agent it'd just hang I guess, but then eventually properly catch up? Kinda weird behavior here, I'd expect it to have a default.

Reply:
See new implementation, should be more robust